File: src/main/java/arc/arc/hooks/lands/LandsMessager.java
package arc.arc.hooks.lands;

import arc.arc.ARC;
import arc.arc.hooks.HookRegistry;
import arc.arc.network.ChannelListener;
import arc.arc.network.RedisManager;
import arc.arc.network.RedisSerializer;
import arc.arc.network.ServerLocation;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;

@RequiredArgsConstructor
public class LandsMessager implements ChannelListener {

    private final RedisManager redisManager;

    @Getter
    private final String reqChannel;

    @Getter
    private final String  respChannel;
    private static BukkitTask cleanupTask;
    Map<UUID, TimedRequest> futures = new ConcurrentHashMap<>();


    public void init(){
        if(cleanupTask != null && cleanupTask.isCancelled()) cleanupTask.cancel();
        cleanupTask = new BukkitRunnable() {
            @Override
            public void run() {
                futures.entrySet()
                        .removeIf(e -> Duration.between(e.getValue().instant(), Instant.now()).getSeconds() > 5);
            }
        }.runTaskTimerAsynchronously(ARC.plugin, 20L, 20L);
    }


    @Override
    public void consume(String channel, String message, String server) {
        if(channel.equals(respChannel)) {
            if (futures.isEmpty()) return;
            LandsRequest landsRequest = RedisSerializer.fromJson(message, LandsRequest.class);
            if (landsRequest == null) return;
            TimedRequest request = futures.get(landsRequest.uuid);
            request.future.complete(landsRequest.serverLocation);
        } else if(channel.equals(reqChannel)){
            if(HookRegistry.landsHook == null) return;
            LandsRequest landsRequest = RedisSerializer.fromJson(message, LandsRequest.class);
            if(landsRequest == null || landsRequest.playerUuid == null) return;
            HookRegistry.landsHook.getSpawnLocation(landsRequest.playerUuid)
                    .thenApply(ServerLocation::of)
                    .thenApply(loc -> new LandsRequest(landsRequest.uuid, landsRequest.playerUuid, loc))
                    .thenApply(RedisSerializer::toJson)
                    .thenAccept(json -> redisManager.publish(respChannel, json));
        }
    }


    public CompletableFuture<ServerLocation> getSpawnLocation(UUID playerUuid){
        UUID uuid = UUID.randomUUID();
        LandsRequest landsRequest = new LandsRequest(uuid, playerUuid, null);
        CompletableFuture<ServerLocation> future = new CompletableFuture<>();
        futures.put(uuid, new TimedRequest(future, Instant.now()));
        redisManager.publish(reqChannel, RedisSerializer.toJson(landsRequest));
        return future;
    }

    record TimedRequest(CompletableFuture<ServerLocation> future, Instant instant){}
}

File: src/main/kotlin/ru/arc/hooks/lands/LandsMessager.kt
package ru.arc.hooks.lands

import ru.arc.common.ServerLocation
import ru.arc.core.ScheduledTask
import ru.arc.core.repeatingAsync
import ru.arc.core.ticks
import ru.arc.hooks.HookRegistry
import ru.arc.redis.ChannelListener
import ru.arc.redis.RedisOperations
import ru.arc.network.RedisSerializer
import java.time.Duration
import java.time.Instant
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeoutException

class LandsMessager(
    private val redis: RedisOperations,
    val reqChannel: String,
    val respChannel: String,
) : ChannelListener {

    val futures: MutableMap<UUID, TimedRequest> = ConcurrentHashMap()
    private var cleanupTask: ScheduledTask? = null

    fun init() {
        close()
        cleanupTask =
            repeatingAsync(20.ticks, delay = 20.ticks) {
                futures.entries.forEach { (id, request) ->
                    val expired = Duration.between(request.instant, Instant.now()).seconds > 5
                    if (expired && futures.remove(id, request)) {
                        request.future.completeExceptionally(TimeoutException("Lands request timed out"))
                    }
                }
            }
    }

    fun close() {
        cleanupTask?.cancel()
        cleanupTask = null
        futures.values.forEach { it.future.cancel(true) }
        futures.clear()
    }

    override fun consume(channel: String, message: String, originServer: String) {
        if (channel == respChannel) {
            if (futures.isEmpty()) return
            val req = RedisSerializer.fromJson(message, LandsRequest::class.java) ?: return
            futures[req.uuid]?.future?.complete(req.serverLocation)
        } else if (channel == reqChannel) {
            if (HookRegistry.landsHook == null) return
            val req = RedisSerializer.fromJson(message, LandsRequest::class.java) ?: return
            val playerUuid = req.playerUuid ?: return
            HookRegistry.landsHook!!.getSpawnLocation(playerUuid)
                .thenApply { ServerLocation.of(it) }
                .thenApply { loc -> LandsRequest(req.uuid, playerUuid, loc) }
                .thenApply { RedisSerializer.toJson(it) }
                .thenAccept { json -> redis.publish(respChannel, json) }
        }
    }

    fun getSpawnLocation(playerUuid: UUID): CompletableFuture<ServerLocation> {
        val uuid = UUID.randomUUID()
        val future = CompletableFuture<ServerLocation>()
        futures[uuid] = TimedRequest(future, Instant.now())
        try {
            redis.publish(reqChannel, RedisSerializer.toJson(LandsRequest(uuid, playerUuid, null)))
        } catch (error: Exception) {
            futures.remove(uuid)
            future.completeExceptionally(error)
        }
        return future
    }

    data class TimedRequest(val future: CompletableFuture<ServerLocation>, val instant: Instant)
}
